File: Cpp/075.cpp
#include<cstdio>
#include<iostream>
#include<algorithm>
#include<string>
#include<vector>
#include<cmath>
#define MAX 1500000
#define DEBUG(x) printf(#x"=%d\n",x)
using namespace std;

int arr[2*MAX];

int gcd(int a, int b)
{
    while(b) b^=a^=b^=a%=b;
    return a;
}

int main()
{
    int m,n,a,b,c,p;
    int ans = 0;
    int upperbound = sqrt(MAX/2);
    for(m=2; m<upperbound; m++)
    {
        for(n=1; n<m; n++)
        {
            if((gcd(m,n)==1 && (m-n)%2==1))
            {
                a = m*m - n*n;
                b = 2*m*n;
                c = m*m + n*n;
                p = a+b+c;
                while(p<=MAX)
                {
                    if(arr[p]==0 || arr[p]==1) arr[p]++;
                    p+=(a+b+c);
                }
            }
        }
    }
    for(int i=0; i<=MAX; i++) if(arr[i] == 1) ans++;
    cout<<ans;
    return 0;
}

File: Ruby/075.rb
# Using Euclid's formula to generate pythogorean triplets (https://en.wikipedia.org/wiki/Pythagorean_triple)
require_relative 'euler'

MAX = 1_500_000
MAX_ROOT = Math.sqrt(MAX)

ans = Hash.new(0)
res = 0
# m > n > 0
1.upto(MAX_ROOT - 1).each do |n|
  (n + 1).upto(MAX_ROOT).each do |m|
    next unless gcd(m, n) == 1 && (m - n) % 2 == 1
    k = 1
    loop do
      p = 2 * m * (m + n) * k
      break if p > MAX
      # puts "a = #{k*(m*m-n*n)}, b = #{k*2*m*n}, c = #{k*(m*m+n*n)}, p = #{p}"
      # puts "m = #{m}, n = #{n}, k = #{k}, p = #{p}"
      ans[p] += 1
      k += 1
    end
  end
end

puts ans.select { |_, v| v == 1}.count
# puts ans.keys
